File: fs/src/main/scala/farjs/fs/popups/DrivePopup.scala
package farjs.fs.popups

import farjs.filelist.FileListState
import farjs.filelist.stack.WithStacks
import farjs.fs.{FSDisk, FSService}
import farjs.ui.border.SingleChars
import farjs.ui.menu.{MenuPopup, MenuPopupProps}
import farjs.ui.task.{Task, TaskAction}
import farjs.ui.{WithSize, WithSizeProps}
import scommons.nodejs.Process.Platform
import scommons.nodejs.process
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

object DrivePopup extends FunctionComponent[DrivePopupProps] {

  private[popups] var platform: Platform = process.platform
  private[popups] var fsService: FSService = FSService.instance
  private[popups] var withSizeComp: ReactClass = WithSize
  private[popups] var menuPopup: ReactClass = MenuPopup

  protected def render(compProps: Props): ReactElement = {
    val (disks, setDisks) = useState(List.empty[FSDisk])
    val props = compProps.plain
    val stacks = WithStacks.useStacks()
    val data = getData(platform, disks)

    val (panelInput, currStack, otherStack) =
      if (props.showOnLeft) (stacks.left.input, stacks.left.stack, stacks.right.stack)
      else (stacks.right.input, stacks.right.stack, stacks.left.stack)
    
    def onAction(dir: String): Unit = {
      val targetDir =
        otherStack.peekLast[FileListState]().state.collect {
          case s if s.currDir.path.startsWith(dir) => s.currDir.path
        }.orElse(
          currStack.peekLast[FileListState]().state.collect {
            case s if s.currDir.path.startsWith(dir) => s.currDir.path
          }
        ).getOrElse(dir)
      
      props.onChangeDir(targetDir)
    }

    useLayoutEffect({ () =>
      val disksF = fsService.readDisks().toFuture.map { disks =>
        setDisks(disks.toList)
      }
      props.dispatch(TaskAction(Task("Reading disks", disksF)))
      ()
    }, Nil)

    if (data.isEmpty) null
    else {
      <(withSizeComp)(^.plain := WithSizeProps { (_, _) =>
        <(menuPopup)(^.plain := MenuPopupProps(
          title = "Drive",
          items = js.Array(data.map(_._2): _*),
          getLeft = { width =>
            val panelWidth =
              if (panelInput != null) panelInput.width
              else 0
  
            MenuPopup.getLeftPos(panelWidth, props.showOnLeft, width)
          },
          onSelect = { index =>
            onAction(data(index)._1)
          },
          onClose = props.onClose
        ))()
      })()
    }
  }

  private val kBytes = 1024d
  private val mBytes = 1024d * kBytes
  private val gBytes = 1024d * mBytes

  private[popups] def getData(platform: Platform, disks: List[FSDisk]): List[(String, String)] = {
    if (disks.isEmpty) Nil
    else {
      val items = disks.map { d =>
        (d.root, d.name, toCompact(d.size), toCompact(d.free))
      }
      val maxNameWidth = 15
      val maxSizeWidth = items.maxBy(_._3.length)._3.length
      val maxFreeWidth = items.maxBy(_._4.length)._4.length
      val sep = SingleChars.vertical

      items.map { case (root, iName, iSize, iFree) =>
        val name = iName.take(maxNameWidth).padTo(maxNameWidth, ' ')
        val size = s"${" " * (maxSizeWidth - iSize.length)}$iSize"
        val free = s"${" " * (maxFreeWidth - iFree.length)}$iFree"
        
        if (platform == Platform.win32) (root, s"  $root $sep$name$sep$size$sep$free ")
        else (root, s" $name$sep$size$sep$free ")
      }
    }
  }
  
  private[popups] def toCompact(bytes: Double): String = {
    if (bytes == 0.0) ""
    else {
      val (size, mod) =
        if (bytes > 1000d * gBytes) (bytes / gBytes, " G")
        else if (bytes > 1000d * mBytes) (bytes / mBytes, " M")
        else if (bytes > 1000d * kBytes) (bytes / kBytes, " K")
        else (bytes, "")
  
      f"$size%.0f$mod"
    }
  }
}

File: fs/popups/DrivePopup.mjs
/**
 * @import PanelStackItem from "@farjs/filelist/stack/PanelStackItem.mjs"
 * @import { Dispatch } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListState } from "@farjs/filelist/FileListState.mjs"
 * @import { FSDisk } from "../FSDisk.mjs"
 * @typedef {import("../FSService.mjs").FSService} FSService
 */
import React, { useLayoutEffect, useState } from "react";
import Task from "@farjs/ui/task/Task.mjs";
import TaskAction from "@farjs/ui/task/TaskAction.mjs";
import MenuPopup from "@farjs/ui/menu/MenuPopup.mjs";
import SingleChars from "@farjs/ui/border/SingleChars.mjs";
import WithSize from "@farjs/ui/WithSize.mjs";
import WithStacks from "@farjs/filelist/stack/WithStacks.mjs";
import FSService from "../FSService.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly dispatch: Dispatch;
 *  readonly showOnLeft: boolean;
 *  onChangeDir(dir: string): void;
 *  onClose(): void;
 * }} DrivePopupProps
 */

/**
 * @param {DrivePopupProps} props
 */
const DrivePopup = (props) => {
  const { platform, fsService, withSizeComp, menuPopup } = DrivePopup;

  const [disks, setDisks] = useState(/** @type {readonly FSDisk[]} */ ([]));

  const stacks = WithStacks.useStacks();
  const data = getData(platform, disks);

  const [panelInput, currStack, otherStack] = props.showOnLeft
    ? [stacks.left.input, stacks.left.stack, stacks.right.stack]
    : [stacks.right.input, stacks.right.stack, stacks.left.stack];

  /** @type {(dir: string) => void} */
  function onAction(dir) {
    /**
     * @returns {string}
     */
    function getTargetDir() {
      /** @type {PanelStackItem<FileListState>} */
      const otherStackItem = otherStack.peekLast();
      const otherState = otherStackItem.state;
      if (otherState && otherState.currDir.path.startsWith(dir)) {
        return otherState.currDir.path;
      }

      /** @type {PanelStackItem<FileListState>} */
      const currStackItem = currStack.peekLast();
      const currState = currStackItem.state;
      if (currState && currState.currDir.path.startsWith(dir)) {
        return currState.currDir.path;
      }

      return dir;
    }

    props.onChangeDir(getTargetDir());
  }

  useLayoutEffect(() => {
    const disksP = fsService.readDisks().then((disks) => {
      setDisks(disks);
    });
    props.dispatch(TaskAction(Task("Reading disks", disksP)));
  }, []);

  return data.length > 0
    ? h(withSizeComp, {
        render: () => {
          return h(menuPopup, {
            title: "Drive",
            items: data.map((_) => _.item),
            getLeft: (width) => {
              const panelWidth =
                panelInput && panelInput.width
                  ? /** @type {number} */ (panelInput.width)
                  : 0;

              return MenuPopup.getLeftPos(panelWidth, props.showOnLeft, width);
            },
            onSelect: (index) => {
              onAction(data[index].dir);
            },
            onClose: props.onClose,
          });
        },
      })
    : null;
};

DrivePopup.displayName = "DrivePopup";
/** @type {NodeJS.Platform} */
DrivePopup.platform = process.platform;
/** @type {FSService} */
DrivePopup.fsService = FSService.instance;
DrivePopup.withSizeComp = WithSize;
DrivePopup.menuPopup = MenuPopup;

/**
 * @param {NodeJS.Platform} platform
 * @param {readonly FSDisk[]} disks
 * @returns {readonly {
 *  readonly dir: string;
 *  readonly item: string;
 * }[]}
 */
function getData(platform, disks) {
  if (disks.length === 0) {
    return [];
  }

  /**
   * @type {readonly {
   *  readonly dir: string;
   *  readonly name: string;
   *  readonly size: string;
   *  readonly free: string;
   * }[]}
   */
  const items = disks.map((d) => {
    return {
      dir: d.root,
      name: d.name,
      size: DrivePopup._toCompact(d.size),
      free: DrivePopup._toCompact(d.free),
    };
  });
  const maxNameWidth = 15;
  const maxSizeWidth = items
    .map((_) => _.size.length)
    .reduce((_1, _2) => (_1 > _2 ? _1 : _2), 0);
  const maxFreeWidth = items
    .map((_) => _.free.length)
    .reduce((_1, _2) => (_1 > _2 ? _1 : _2), 0);
  const sep = SingleChars.vertical;

  return items.map(({ dir, name: iName, size: iSize, free: iFree }) => {
    const name = iName
      .slice(0, Math.min(maxNameWidth, iName.length))
      .padEnd(maxNameWidth);
    const size = `${" ".repeat(maxSizeWidth - iSize.length)}${iSize}`;
    const free = `${" ".repeat(maxFreeWidth - iFree.length)}${iFree}`;

    return platform === "win32"
      ? {
          dir,
          item: `  ${dir} ${sep}${name}${sep}${size}${sep}${free} `,
        }
      : { dir, item: ` ${name}${sep}${size}${sep}${free} ` };
  });
}

const kBytes = 1024;
const mBytes = 1024 * kBytes;
const gBytes = 1024 * mBytes;

/**
 * @type {(bytes: number) => string}
 */
DrivePopup._toCompact = (bytes) => {
  if (bytes === 0) {
    return "";
  }

  const [size, mod] = (() => {
    return bytes > 1000 * gBytes
      ? [bytes / gBytes, " G"]
      : bytes > 1000 * mBytes
      ? [bytes / mBytes, " M"]
      : bytes > 1000 * kBytes
      ? [bytes / kBytes, " K"]
      : [bytes, ""];
  })();

  return `${Math.round(size)}${mod}`;
};

export default DrivePopup;
